sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,311 @@
+                          leaf_size=leaf_size, metric=metric, p=p,
+                          metric_params=metric_params, n_jobs=n_jobs)
+
+        self.get_n_jobs_ = _get_n_jobs(self.n_jobs)
+        self.contamination = contamination
+
+    def fit(self, X, y=None):
+        """Fit the model using X as training data
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix, BallTree, KDTree}
+            Training data. If array or matrix, shape [n_samples, n_features],
+            or [n_samples, n_samples] if metric='precomputed'.
+        """
+        super(LocalOutlierFactor, self).fit(X)
+
+        # Compute decision_function over training samples to define threshold_:
+        self.decision_function_train_ = self.decision_function()
+
+        self.threshold_ = -sp.stats.scoreatpercentile(
+            -self.decision_function_train_, 100. * (1. - self.contamination))
+        return self
+
+    def predict(self, X=None):
+        """Predict Local Outlier Factor of X.
+
+        The local outlier factor (LOF) of a sample captures its
+        supposed `degree of abnormality'.
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features), default=None
+            The query sample or samples to compute the Local Outlier Factor
+            wrt to the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        is_inlier : array of shape (n_samples,)
+            Returns 1 for anomalies/outliers and -1 for inliers.
+        """
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
+            is_inlier = np.ones(X.shape[0], dtype=int)
+        else:
+            is_inlier = np.ones(self._fit_X.shape[0], dtype=int)
+
+        is_inlier[self.decision_function(X) <= self.threshold_] = -1
+
+        return is_inlier
+
+    def decision_function(self, X=None):
+        """Opposite of the Local Outlier Factor of X (as bigger is better).
+
+        The local outlier factor (LOF) of a sample captures its
+        supposed `degree of abnormality'.
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features), optional
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        lof_scores : array of shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
```

Comment:
You have to be sure that you've done `fit` (see the comment for the predict function)